File: MiniDOMOffice/Tickets/CTicketCursor.vb
Imports minidom
Imports minidom.Databases
Imports minidom.Sistema

Partial Public Class Office

    ''' <summary>
    ''' Cursore sulla tabelle delle segnalazioni
    ''' </summary>
    ''' <remarks></remarks>
    Public NotInheritable Class CTicketCursor
        Inherits DBObjectCursorPO(Of CTicket)

        Private m_IDApertoDa As New CCursorField(Of Integer)("ApertoDa")
        Private m_NomeApertoDa As New CCursorFieldObj(Of String)("NomeApertoDa")

        Private m_IDCliente As New CCursorField(Of Integer)("IDCliente")
        Private m_NomeCliente As New CCursorFieldObj(Of String)("NomeCliente")

        Private m_IDInCaricoA As New CCursorField(Of Integer)("InCaricoA")
        Private m_NomeInCaricoA As New CCursorFieldObj(Of String)("InCaricoANome")
        Private m_Categoria As New CCursorFieldObj(Of String)("Categoria")
        Private m_Sottocategoria As New CCursorFieldObj(Of String)("Sottocategoria")
        Private m_Messaggio As New CCursorFieldObj(Of String)("Messaggio")
        Private m_StatoSegnalazione As New CCursorField(Of TicketStatus)("StatoSegnalazione")
        Private m_Priorita As New CCursorField(Of PriorityEnum)("Priorita")
        Private m_IDSupervisore As New CCursorField(Of Integer)("IDSupervisore")
        Private m_NomeSupervisore As New CCursorFieldObj(Of String)("NomeSupervisore")
        Private m_Canale As New CCursorFieldObj(Of String)("Canale")
        Private m_DataRichiesta As New CCursorField(Of Date)("DataRichiesta")
        Private m_DataPresaInCarico As New CCursorField(Of Date)("DataPresaInCarico")
        Private m_DataChiusura As New CCursorField(Of Date)("DataChiusura")

        Private m_IDContesto As New CCursorField(Of Integer)("IDContesto")
        Private m_TipoContesto As New CCursorFieldObj(Of String)("TipoContesto")

        Private m_IDPostazione As New CCursorField(Of Integer)("IDPostazione")
        Private m_NomePostazione As New CCursorFieldObj(Of String)("NomePostazione")

        Public Sub New()
        End Sub

        Public ReadOnly Property IDPostazione As CCursorField(Of Integer)
            Get
                Return Me.m_IDPostazione
            End Get
        End Property

        Public ReadOnly Property NomePostazione As CCursorFieldObj(Of String)
            Get
                Return Me.m_NomePostazione
            End Get
        End Property


        Public ReadOnly Property IDContesto As CCursorField(Of Integer)
            Get
                Return Me.m_IDContesto
            End Get
        End Property

        Public ReadOnly Property TipoContesto As CCursorFieldObj(Of String)
            Get
                Return Me.m_TipoContesto
            End Get
        End Property


        Public ReadOnly Property IDCliente As CCursorField(Of Integer)
            Get
                Return Me.m_IDCliente
            End Get
        End Property

        Public ReadOnly Property NomeCliente As CCursorFieldObj(Of String)
            Get
                Return Me.m_NomeCliente
            End Get
        End Property


        Public ReadOnly Property IDApertoDa As CCursorField(Of Integer)
            Get
                Return Me.m_IDApertoDa
            End Get
        End Property

        Public ReadOnly Property NomeApertoDa As CCursorFieldObj(Of String)
            Get
                Return Me.m_NomeApertoDa
            End Get
        End Property

        Public ReadOnly Property IDInCaricoA As CCursorField(Of Integer)
            Get
                Return Me.m_IDInCaricoA
            End Get
        End Property

        Public ReadOnly Property NomeInCaricoA As CCursorFieldObj(Of String)
            Get
                Return Me.m_NomeInCaricoA
            End Get
        End Property

        Public ReadOnly Property IDSupervisore As CCursorField(Of Integer)
            Get
                Return Me.m_IDSupervisore
            End Get
        End Property

        Public ReadOnly Property NomeSupervisore As CCursorFieldObj(Of String)
            Get
                Return Me.m_NomeSupervisore
            End Get
        End Property

        Public ReadOnly Property Categoria As CCursorFieldObj(Of String)
            Get
                Return Me.m_Categoria
            End Get
        End Property

        Public ReadOnly Property Sottocategoria As CCursorFieldObj(Of String)
            Get
                Return Me.m_Sottocategoria
            End Get
        End Property

        Public ReadOnly Property Canale As CCursorFieldObj(Of String)
            Get
                Return Me.m_Canale
            End Get
        End Property

        Public ReadOnly Property Messaggio As CCursorFieldObj(Of String)
            Get
                Return Me.m_Messaggio
            End Get
        End Property

        Public ReadOnly Property StatoSegnalazione As CCursorField(Of TicketStatus)
            Get
                Return Me.m_StatoSegnalazione
            End Get
        End Property

        Public ReadOnly Property DataRichiesta As CCursorField(Of Date)
            Get
                Return Me.m_DataRichiesta
            End Get
        End Property

        Public ReadOnly Property DataPresaInCarico As CCursorField(Of Date)
            Get
                Return Me.m_DataPresaInCarico
            End Get
        End Property

        Public ReadOnly Property DataChiusura As CCursorField(Of Date)
            Get
                Return Me.m_DataChiusura
            End Get
        End Property

        Public ReadOnly Property Priorita As CCursorField(Of PriorityEnum)
            Get
                Return Me.m_Priorita
            End Get
        End Property


        Public Overrides Function GetTableName() As String
            Return "tbl_SupportTickets"
        End Function

        Protected Overrides Function GetConnection() As CDBConnection
            Return Office.Database
        End Function

        Protected Overrides Function GetModule() As CModule
            Return Tickets.Module
        End Function

        Public Overrides Function Add() As Object
            Dim ret As CTicket = MyBase.Add()
            ret.ApertoDa = Users.CurrentUser
            ret.DataRichiesta = Now
            Return ret
        End Function

        Public Overrides Function GetWherePartLimit() As String
            If Not Me.Module.UserCanDoAction("list") AndAlso Me.Module.UserCanDoAction("list_category") Then
                Dim categories As CCollection(Of CTicketCategory) = TicketCategories.GetUserAllowedCategories(Sistema.Users.CurrentUser)
                If (categories.Count = 0) Then Return MyBase.GetWherePartLimit()

                Dim buff As New System.Text.StringBuilder
                For Each cat As CTicketCategory In categories
                    If (buff.Length > 0) Then buff.Append(" OR ")
                    If (cat.Sottocategoria = "") Then
                        buff.Append("([Categoria]=" & DBUtils.DBString(cat.Categoria) & ")")
                    Else
                        buff.Append("([Categoria]=" & DBUtils.DBString(cat.Categoria) & " AND [Sottocategoria]=" & DBUtils.DBString(cat.Sottocategoria) & ")")
                    End If
                Next

                Return Strings.Combine("(" & MyBase.GetWherePartLimit & ")", "(" & buff.ToString & ")", " OR ")
            Else
                Return MyBase.GetWherePartLimit
            End If
        End Function

    End Class


End Class

File: MiniDOMOffice/Tickets/CTicketCursor.cs
using System;
using System.Collections;
using System.Collections.Generic;
using DMD;
using DMD.XML;
using DMD.Databases;
using DMD.Databases.Collections;
using minidom;
using minidom.repositories;
using static minidom.Sistema;
using static minidom.Anagrafica;
using static minidom.Office;


namespace minidom
{
    public partial class Office
    {

        /// <summary>
        /// Cursore di <see cref="CTicket"/>
        /// </summary>
        /// <remarks></remarks>
        [Serializable]
        public sealed class CTicketCursor
            : minidom.Databases.DBObjectCursorPO<CTicket>
        {
            private DBCursorField<int> m_IDApertoDa = new DBCursorField<int>("ApertoDa");
            private DBCursorStringField m_NomeApertoDa = new DBCursorStringField("NomeApertoDa");
            private DBCursorField<int> m_IDCliente = new DBCursorField<int>("IDCliente");
            private DBCursorStringField m_NomeCliente = new DBCursorStringField("NomeCliente");
            private DBCursorField<int> m_IDInCaricoA = new DBCursorField<int>("InCaricoA");
            private DBCursorStringField m_NomeInCaricoA = new DBCursorStringField("InCaricoANome");
            private DBCursorStringField m_Categoria = new DBCursorStringField("Categoria");
            private DBCursorStringField m_Sottocategoria = new DBCursorStringField("Sottocategoria");
            private DBCursorStringField m_Messaggio = new DBCursorStringField("Messaggio");
            private DBCursorField<TicketStatus> m_StatoSegnalazione = new DBCursorField<TicketStatus>("StatoSegnalazione");
            private DBCursorField<PriorityEnum> m_Priorita = new DBCursorField<PriorityEnum>("Priorita");
            private DBCursorField<int> m_IDSupervisore = new DBCursorField<int>("IDSupervisore");
            private DBCursorStringField m_NomeSupervisore = new DBCursorStringField("NomeSupervisore");
            private DBCursorStringField m_Canale = new DBCursorStringField("Canale");
            private DBCursorField<DateTime> m_DataRichiesta = new DBCursorField<DateTime>("DataRichiesta");
            private DBCursorField<DateTime> m_DataPresaInCarico = new DBCursorField<DateTime>("DataPresaInCarico");
            private DBCursorField<DateTime> m_DataChiusura = new DBCursorField<DateTime>("DataChiusura");
            private DBCursorField<int> m_IDContesto = new DBCursorField<int>("IDContesto");
            private DBCursorStringField m_TipoContesto = new DBCursorStringField("TipoContesto");
            private DBCursorField<int> m_IDPostazione = new DBCursorField<int>("IDPostazione");
            private DBCursorStringField m_NomePostazione = new DBCursorStringField("NomePostazione");

            /// <summary>
            /// Costruttore
            /// </summary>
            public CTicketCursor()
            {
            }

            /// <summary>
            /// IDPostazione
            /// </summary>
            public DBCursorField<int> IDPostazione
            {
                get
                {
                    return m_IDPostazione;
                }
            }

            /// <summary>
            /// NomePostazione
            /// </summary>
            public DBCursorStringField NomePostazione
            {
                get
                {
                    return m_NomePostazione;
                }
            }

            /// <summary>
            /// IDContesto
            /// </summary>
            public DBCursorField<int> IDContesto
            {
                get
                {
                    return m_IDContesto;
                }
            }

            /// <summary>
            /// TipoContesto
            /// </summary>
            public DBCursorStringField TipoContesto
            {
                get
                {
                    return m_TipoContesto;
                }
            }

            /// <summary>
            /// IDCliente
            /// </summary>
            public DBCursorField<int> IDCliente
            {
                get
                {
                    return m_IDCliente;
                }
            }

            /// <summary>
            /// NomeCliente
            /// </summary>
            public DBCursorStringField NomeCliente
            {
                get
                {
                    return m_NomeCliente;
                }
            }

            /// <summary>
            /// IDApertoDa
            /// </summary>
            public DBCursorField<int> IDApertoDa
            {
                get
                {
                    return m_IDApertoDa;
                }
            }

            /// <summary>
            /// NomeApertoDa
            /// </summary>
            public DBCursorStringField NomeApertoDa
            {
                get
                {
                    return m_NomeApertoDa;
                }
            }

            /// <summary>
            /// IDInCaricoA
            /// </summary>
            public DBCursorField<int> IDInCaricoA
            {
                get
                {
                    return m_IDInCaricoA;
                }
            }

            /// <summary>
            /// NomeInCaricoA
            /// </summary>
            public DBCursorStringField NomeInCaricoA
            {
                get
                {
                    return m_NomeInCaricoA;
                }
            }

            /// <summary>
            /// IDSupervisore
            /// </summary>
            public DBCursorField<int> IDSupervisore
            {
                get
                {
                    return m_IDSupervisore;
                }
            }

            /// <summary>
            /// NomeSupervisore
            /// </summary>
            public DBCursorStringField NomeSupervisore
            {
                get
                {
                    return m_NomeSupervisore;
                }
            }

            /// <summary>
            /// Categoria
            /// </summary>
            public DBCursorStringField Categoria
            {
                get
                {
                    return m_Categoria;
                }
            }

            /// <summary>
            /// Sottocategoria
            /// </summary>
            public DBCursorStringField Sottocategoria
            {
                get
                {
                    return m_Sottocategoria;
                }
            }

            /// <summary>
            /// Canale
            /// </summary>
            public DBCursorStringField Canale
            {
                get
                {
                    return m_Canale;
                }
            }

            /// <summary>
            /// Messaggio
            /// </summary>
            public DBCursorStringField Messaggio
            {
                get
                {
                    return m_Messaggio;
                }
            }

            /// <summary>
            /// StatoSegnalazione
            /// </summary>
            public DBCursorField<TicketStatus> StatoSegnalazione
            {
                get
                {
                    return m_StatoSegnalazione;
                }
            }

            /// <summary>
            /// DataRichiesta
            /// </summary>
            public DBCursorField<DateTime> DataRichiesta
            {
                get
                {
                    return m_DataRichiesta;
                }
            }

            /// <summary>
            /// DataPresaInCarico
            /// </summary>
            public DBCursorField<DateTime> DataPresaInCarico
            {
                get
                {
                    return m_DataPresaInCarico;
                }
            }

            /// <summary>
            /// DataChiusura
            /// </summary>
            public DBCursorField<DateTime> DataChiusura
            {
                get
                {
                    return m_DataChiusura;
                }
            }

            /// <summary>
            /// Priorita
            /// </summary>
            public DBCursorField<PriorityEnum> Priorita
            {
                get
                {
                    return m_Priorita;
                }
            }
  

            /// <summary>
            /// Inizializza i parametri per un nuovo oggetto
            /// </summary>
            /// <param name="item"></param>
            protected override void OnInitialize(CTicket item)
            {
                item.ApertoDa = Sistema.Users.CurrentUser;
                item.DataRichiesta = DMD.DateUtils.Now();
                base.OnInitialize(item);
            }

            /// <summary>
            /// Prepara i vincoli relativi alla sicurezza
            /// </summary>
            /// <returns></returns>
            protected override DBCursorFieldBase GetWherePartLimit()
            {
                if (!Module.UserCanDoAction("list") && Module.UserCanDoAction("list_category"))
                {
                    var categories = minidom.Office.TicketCategories.GetUserAllowedCategories(Sistema.Users.CurrentUser);
                    if (categories.Count == 0)
                        return base.GetWherePartLimit();
                    var buff = DBCursorField.False;
                    foreach (var cat in categories)
                    {
                        if (string.IsNullOrEmpty(cat.Sottocategoria))
                        {
                            buff += this.Field("Categoria").EQ(cat.Categoria);
                        }
                        else
                        {
                            buff += this.Field("Categoria").EQ(cat.Categoria) * this.Field("Sottocategoria").EQ(cat.Sottocategoria);
                        }
                    }

                    return base.GetWherePartLimit() + buff;
                }
                else
                {
                    return base.GetWherePartLimit();
                }
            }
        }
    }
}